File: py-standalone/core/pylib/stores/kv.py
"""
KV-backed StepMachineStore for Python step-machine.

Parallels JS KVStorageStore key schema:
  state_<b64url(runId)>
  data_<b64url(runId)>_<b64url(key)>
"""
from __future__ import annotations

import base64
import json
from pathlib import Path
from typing import Any, Dict, List


def _b64url_encode(value: str) -> str:
    return base64.urlsafe_b64encode(value.encode("utf-8")).decode("ascii").rstrip("=")


def _b64url_decode(value: str) -> str:
    padded = value + ("=" * ((4 - (len(value) % 4)) % 4))
    return base64.urlsafe_b64decode(padded.encode("ascii")).decode("utf-8")


class FsKvStorage:
    """Simple file-backed KV storage where each key maps to <root>/<key>.json."""

    def __init__(self, root_dir: str):
        self._root = Path(root_dir)
        self._root.mkdir(parents=True, exist_ok=True)

    def _path_for_key(self, key: str) -> Path:
        if not key:
            raise ValueError("KV key must be non-empty")
        return self._root / f"{key}.json"

    def read(self, key: str) -> Any:
        p = self._path_for_key(key)
        if not p.exists():
            return None
        with p.open("r", encoding="utf-8") as f:
            return json.load(f)

    def write(self, key: str, value: Any) -> None:
        p = self._path_for_key(key)
        p.parent.mkdir(parents=True, exist_ok=True)
        with p.open("w", encoding="utf-8") as f:
            json.dump(value, f, indent=2, ensure_ascii=True)

    def delete(self, key: str) -> None:
        p = self._path_for_key(key)
        if p.exists():
            p.unlink()

    def list_keys(self, prefix: str = "") -> List[str]:
        keys: List[str] = []
        for p in self._root.glob("*.json"):
            key = p.stem
            if prefix and not key.startswith(prefix):
                continue
            keys.append(key)
        keys.sort()
        return keys


class KVStorageStore:
    def __init__(self, kv: Any):
        self._kv = kv

    def _state_key(self, run_id: str) -> str:
        return f"state_{_b64url_encode(run_id)}"

    def _data_prefix(self, run_id: str) -> str:
        return f"data_{_b64url_encode(run_id)}_"

    def _data_key(self, run_id: str, key: str) -> str:
        return f"{self._data_prefix(run_id)}{_b64url_encode(key)}"

    def save_run_state(self, run_id: str, state: Dict[str, Any]) -> None:
        self._kv.write(self._state_key(run_id), state)

    def load_run_state(self, run_id: str) -> Dict[str, Any] | None:
        value = self._kv.read(self._state_key(run_id))
        return value if isinstance(value, dict) else None

    def delete_run_state(self, run_id: str) -> None:
        self._kv.delete(self._state_key(run_id))
        for key in self._kv.list_keys(self._data_prefix(run_id)):
            self._kv.delete(key)

    def set_data(self, run_id: str, key: str, value: Any) -> None:
        self._kv.write(self._data_key(run_id, key), value)

    def get_data(self, run_id: str, key: str) -> Any:
        return self._kv.read(self._data_key(run_id, key))

    def get_all_data(self, run_id: str) -> Dict[str, Any]:
        prefix = self._data_prefix(run_id)
        out: Dict[str, Any] = {}
        for key in self._kv.list_keys(prefix):
            decoded = _b64url_decode(key[len(prefix):])
            out[decoded] = self._kv.read(key)
        return out

    def clear_data(self, run_id: str) -> None:
        for key in self._kv.list_keys(self._data_prefix(run_id)):
            self._kv.delete(key)

    def list_runs(self) -> List[str]:
        runs: List[str] = []
        for key in self._kv.list_keys("state_"):
            encoded = key[len("state_"):]
            runs.append(_b64url_decode(encoded))
        return runs

File: src/stores/kv.ts
/**
 * yaml-flow - KVStorageStore
 *
 * StepMachineStore backed by a platform-agnostic KVStorage.
 * The platform adapter provides the right KVStorage implementation
 * (Node.js FS, browser localStorage, in-memory, CosmosDB, etc.).
 *
 * Key schema (all parts are base64url-encoded — alphanumeric + '-' + '_' only):
 *   state_<b64(runId)>              → StepMachineState
 *   data_<b64(runId)>_<b64(key)>   → arbitrary value
 */

import type { StepMachineStore, StepMachineState } from '../step-machine/types.js';
import type { KVStorage } from '../cli/common/storage-interface.js';

function b64url(s: string): string {
  return btoa(s).replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '');
}

function fromb64url(s: string): string {
  return atob(s.replace(/-/g, '+').replace(/_/g, '/'));
}

export class KVStorageStore implements StepMachineStore {
  constructor(private readonly kv: KVStorage) {}

  private stateKey(runId: string): string { return `state_${b64url(runId)}`; }
  private dataPrefix(runId: string): string { return `data_${b64url(runId)}_`; }
  private dataKey(runId: string, key: string): string { return `${this.dataPrefix(runId)}${b64url(key)}`; }

  async saveRunState(runId: string, state: StepMachineState): Promise<void> {
    this.kv.write(this.stateKey(runId), state);
  }

  async loadRunState(runId: string): Promise<StepMachineState | null> {
    const v = this.kv.read(this.stateKey(runId));
    return v != null && typeof v === 'object' ? (v as StepMachineState) : null;
  }

  async deleteRunState(runId: string): Promise<void> {
    this.kv.delete(this.stateKey(runId));
    for (const k of this.kv.listKeys(this.dataPrefix(runId))) this.kv.delete(k);
  }

  async setData(runId: string, key: string, value: unknown): Promise<void> {
    this.kv.write(this.dataKey(runId, key), value);
  }

  async getData(runId: string, key: string): Promise<unknown> {
    return this.kv.read(this.dataKey(runId, key));
  }

  async getAllData(runId: string): Promise<Record<string, unknown>> {
    const prefix = this.dataPrefix(runId);
    const result: Record<string, unknown> = {};
    for (const k of this.kv.listKeys(prefix)) {
      result[fromb64url(k.slice(prefix.length))] = this.kv.read(k);
    }
    return result;
  }

  async clearData(runId: string): Promise<void> {
    for (const k of this.kv.listKeys(this.dataPrefix(runId))) this.kv.delete(k);
  }

  async listRuns(): Promise<string[]> {
    return this.kv.listKeys('state_').map(k => fromb64url(k.slice('state_'.length)));
  }
}
